control-plane/p4RuntimeArchHandler.h

```diff
@@ -0,0 +1,384 @@
+        return P4RuntimeSymbolType(::p4::config::v1::P4Ids::VALUE_SET);
+    }
+    static P4RuntimeSymbolType CONTROLLER_HEADER() {
+        return P4RuntimeSymbolType(::p4::config::v1::P4Ids::CONTROLLER_HEADER);
+    }
+
+    bool operator==(const P4RuntimeSymbolType& other) const {
+        return id == other.id;
+    }
+
+    bool operator!=(const P4RuntimeSymbolType& other) const {
+        return !(*this == other);
+    }
+
+    bool operator<(const P4RuntimeSymbolType& other) const {
+        return id < other.id;
+    }
+
+ protected:
+    static P4RuntimeSymbolType make(p4rt_id_t id) {
+        return P4RuntimeSymbolType(id);
+    }
+
+ private:
+    // even if the constructor is protected, the static functions in the derived
+    // classes cannot access it, which is why we use the make factory function
+    constexpr P4RuntimeSymbolType(p4rt_id_t id) noexcept
+        : id(id) { }
+
+    /// The 8-bit id prefix for that type, as per the p4info.proto file.
+    p4rt_id_t id;
+};
+
+/// A table which tracks the symbols which are visible to P4Runtime and their
+/// ids.
+class P4RuntimeSymbolTableIface {
+ public:
+    virtual ~P4RuntimeSymbolTableIface() { }
+    /// Add a @type symbol, extracting the name and id from @declaration.
+    virtual void add(P4RuntimeSymbolType type, const IR::IDeclaration* declaration) = 0;
+    /// Add a @type symbol with @name and possibly an explicit P4 '@id'.
+    virtual void add(P4RuntimeSymbolType type, cstring name,
+                     boost::optional<p4rt_id_t> id = boost::none) = 0;
+    /// @return the P4Runtime id for the symbol of @type corresponding to
+    /// @declaration.
+    virtual p4rt_id_t getId(P4RuntimeSymbolType type,
+                            const IR::IDeclaration* declaration) const = 0;
+    /// @return the P4Runtime id for the symbol of @type with name @name.
+    virtual p4rt_id_t getId(P4RuntimeSymbolType type, cstring name) const = 0;
+    /// @return the alias for the given fully qualified external name. P4Runtime
+    /// defines an alias for each object to make referring to objects easier.
+    /// By default, the alias is the shortest unique suffix of path components
+    /// in the name.
+    virtual cstring getAlias(cstring name) const = 0;
+};
+
+/// The interface for defining the P4Info serialization logic for a specific P4
+/// architecture. The goal is to reduce code duplication between
+/// architectures. The @ref P4RuntimeSerializer will call these methods when
+/// generating the P4Info message to handle architecture-specific parts. @ref
+/// P4RuntimeSerializer generates the P4Info in two passes: first it collects
+/// all the control-plane visible symbols from the program into the symbol
+/// table, then it builds the P4Info message by adding each collected entity to
+/// the Protobuf message. The collect* methods are called in the first pass, the
+/// add* methods are called in the second pass.
+class P4RuntimeArchHandlerIface {
```

Comment:
I would have implemented this as a pass manager with two passes, just to take advantage of the visiting functions, rather than rolling out another visiting infra.